src/subprocess-posix.cc

```diff
@@ -169,11 +170,61 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+bool SubprocessSet::child_exited_;
+
+void SubprocessSet::SetChildExited(int signum) {
+  child_exited_ = true;
+}
+
+void SubprocessSet::HandleChildExit() {
+  if (!child_exited_)
+    return;
+  child_exited_ = false;
+  for (int i = running_.size(); i > 0; --i)
+    WaitChildExit(false);
+}
+
+void SubprocessSet::WaitChildExit(bool block) {
+  int status = 0;
+  int options = block ? 0 : WNOHANG;
+  pid_t pid = waitpid(-1, &status, options);
+  if (pid == -1)
+    Fatal("waitpid: %s", strerror(errno));
+  if (pid == 0)
+    return;
+  for (vector<Subprocess*>::iterator i = running_.begin();
```

Comment:
> This loops kinda duplicates code in DoWork() (the one right after 'nfds_t cur_nfd = 0;')

Kinda, but not exactly. The one in DoWork is responding to known readability, whereas the one in HandleChildExit just blindly reads until there's nothing left. For DoWork to handle both cases, I think it would have to change considerably.

> Wouldn't it be simpler to not have this additional loop here at all, rather to call HandleChildExit() just before the return from DoWork(), after all pipes are drained?

How do we know which pipes to drain if we haven't yet handled child exit? Just do a non-blocking read loop on all FDs after ppoll/pselect returns? That doesn't sound very efficient. And we can't just wait for a read event from ppoll, because then the original issue would not be solved; i.e. the pipe might remain open, but not become readable.